server/src/test/java/org/elasticsearch/cluster/routing/IndexRoutingTests.java

```diff
@@ -714,6 +716,118 @@ public void testRoutingPathLogsdb() throws IOException {
+                    .put(IndexSettings.MODE.getKey(), IndexMode.TIME_SERIES)
+            )
+            .numberOfShards(shards)
+            .numberOfReplicas(0)
+            .setRoutingNumShards(shards)
+            .build();
+        IndexRouting routing = IndexRouting.fromIndexMetadata(startingMetadata);
+        IndexRouting splitRouting = getSplitRouting(startingMetadata);
+
+        int iters = randomIntBetween(100, 1000);
+        for (int i = 0; i < iters; i++) {
+            IndexRequest req = new IndexRequest();
+            routing.preProcess(req);
+            assertNull(req.id());
+
+            // Verify that routing uses the field name and value in the routing path.
+            String value = randomAlphaOfLength(20);
+            int expectedShard = expectedShard(routing, List.of("top", value), shards);
+            req.source(Map.of("top", value, "bar", "B"));
+            assertEquals(expectedShard, routing.indexShard(req));
+
+            validateReroute(routing, req, expectedShard, splitRouting, shards);
+        }
+    }
+
+    public void testRerouteToTargetTsidBeforeWrittenToRouting() throws IOException {
+        int shards = between(2, 500);
+        IndexMetadata startingMetadata = IndexMetadata.builder("test")
+            .settings(
+                settings(IndexVersion.fromId(IndexVersions.TIME_SERIES_ROUTING_HASH_IN_ID.id() - 1)).put(
+                    IndexMetadata.INDEX_ROUTING_PATH.getKey(),
+                    "top"
+                ).put(IndexSettings.MODE.getKey(), IndexMode.TIME_SERIES)
+            )
+            .numberOfShards(shards)
+            .numberOfReplicas(0)
+            .setRoutingNumShards(shards)
+            .build();
+        IndexRouting routing = IndexRouting.fromIndexMetadata(startingMetadata);
+        IndexRouting splitRouting = getSplitRouting(startingMetadata);
+
+        int iters = randomIntBetween(100, 1000);
+        for (int i = 0; i < iters; i++) {
+            IndexRequest req = new IndexRequest();
+            routing.preProcess(req);
+            assertNull(req.id());
+
+            // Verify that routing uses the field name and value in the routing path.
+            String value = randomAlphaOfLength(20);
+            int expectedShard = expectedShard(routing, List.of("top", value), shards);
+            req.source(Map.of("top", value, "bar", "B"));
+            assertEquals(expectedShard, routing.indexShard(req));
+
+            validateReroute(routing, req, expectedShard, splitRouting, shards);
+        }
+    }
+
+    private static void validateReroute(IndexRouting routing, IndexRequest req, int expectedShard, IndexRouting splitRouting, int shards) {
+        routing.postProcess(req);
+        assertEquals(expectedShard, routing.rerouteToTarget(req));
+        assertThat(splitRouting.rerouteToTarget(req), either(equalTo(expectedShard)).or(equalTo(expectedShard + shards)));
```

Comment:
later we might want to add a `targetShard` method to `IndexReshardingState.Split` (similar to the existing `sourceShard` method) to keep the logic for associating source and target shards in one place instead of hand-rolling it.